Source/Fluids/WarpXFluidContainer.cpp

```diff
@@ -0,0 +1,815 @@
+/* Copyright 2023 Grant Johnson, Remi Lehe
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "ablastr/coarsen/sample.H"
+#include "Particles/Pusher/UpdateMomentumHigueraCary.H"
+
+#include "MusclHancockUtils.H"
+#include "WarpXFluidContainer.H"
+#include "WarpX.H"
+#include <ablastr/utils/Communication.H>
+using namespace ablastr::utils::communication;
+using namespace amrex;
+
+WarpXFluidContainer::WarpXFluidContainer(int nlevs_max, int ispecies, const std::string &name)
+{
+    species_id = ispecies;
+    species_name = name;
+
+    plasma_injector = std::make_unique<PlasmaInjector>(species_id, species_name);
+    physical_species = plasma_injector->getPhysicalSpecies();
+    charge = plasma_injector->getCharge();
+    mass = plasma_injector->getMass();
+
+    ReadParameters();
+
+    // Resize the list of MultiFabs for the right number of levels
+    N.resize(nlevs_max);
+    NU.resize(nlevs_max);
+}
+
+void WarpXFluidContainer::ReadParameters()
+{
+    static bool initialized = false;
+    if (!initialized)
+    {
+        const ParmParse pp_species_name(species_name);
+        pp_species_name.query("do_not_deposit", do_not_deposit);
+        pp_species_name.query("do_not_gather", do_not_gather);
+        pp_species_name.query("do_not_push", do_not_push);
+        initialized = true;
+    }
+}
+
+void WarpXFluidContainer::AllocateLevelMFs(int lev, const BoxArray &ba, const DistributionMapping &dm)
+{
+    int ncomps = 1;
+    amrex::IntVect nguards = {AMREX_D_DECL(2, 2, 2)};
+
+    auto &warpx = WarpX::GetInstance();
```

Comment:
Could you replace all the `auto` by the corresponding type (unless it is a very complicated type), throughout this file?
This helps when other people read the code (so that they know which type the objects have).
```suggestion
    WarpX &warpx = WarpX::GetInstance();
```